python/neuroglancer/pipeline/volumes/cloudvolume.py

```diff
@@ -0,0 +1,486 @@
+      delta = min2(end - start, img3d.shape[:3])
+      end = start + delta
+
+      renderbuffer[ start.x:end.x, start.y:end.y, start.z:end.z, : ] = img3d[ :delta.x, :delta.y, :delta.z, : ]
+
+    bounded_request = Bbox.clamp(requested_bbox, self.bounds)
+    lp = bounded_request.minpt - realized_bbox.minpt # low realized point
+    hp = lp + bounded_request.size3()
+
+    renderbuffer = renderbuffer[ lp.x:hp.x:steps.x, lp.y:hp.y:steps.y, lp.z:hp.z:steps.z, channel_slice ] 
+    return VolumeCutout.from_volume(self, renderbuffer, bounded_request)
+  
+  def __setitem__(self, slices, img):
+    imgshape = list(img.shape)
+    if len(imgshape) == 3:
+      imgshape = imgshape + [ self.num_channels ]
+
+    maxsize = list(self.bounds.maxpt) + [ self.num_channels ]
+    minsize = list(self.bounds.minpt) + [ 0 ]
+    slices = generate_slices(slices, minsize, maxsize)
+    bbox = Bbox.from_slices(slices)
+
+    slice_shape = list(bbox.size3()) + [ slices[3].stop - slices[3].start ]
+
+    if not np.array_equal(imgshape, slice_shape):
+      raise ValueError("Illegal slicing, Image shape: {} != {} Slice Shape".format(imgshape, slice_shape))
+    
+    self.upload_image(img, bbox.minpt)
+
+  def upload_image(self, img, offset):
+    if str(self.dtype) != str(img.dtype):
+      raise ValueError('The uploaded image data type must match the volume data type. volume: {}, image: {}'.format(self.dtype, img.dtype))
+
+    uploads = []
+    for imgchunk, spt, ept in tqdm(self._generate_chunks(img, offset), desc='uploading image'):
+      if np.array_equal(spt, ept):
+          continue
+
+      # handle the edge of the dataset
+      clamp_ept = min2(ept, self.bounds.maxpt)
+      newept = clamp_ept - spt
+      imgchunk = imgchunk[ :newept.x, :newept.y, :newept.z, : ]
+
+      filename = "{}-{}_{}-{}_{}-{}".format(
+        spt.x, clamp_ept.x,
+        spt.y, clamp_ept.y, 
+        spt.z, clamp_ept.z
+      )
+
+      cloudpath = os.path.join(self.key, filename)
+      encoded = neuroglancer.chunks.encode(imgchunk, self.encoding)
+      uploads.append( (cloudpath, encoded) )
+
+    content_type = 'application/octet-stream'
+    if self.encoding == 'jpeg':
+      content_type == 'image/jpeg'
+
+    compress = (self.layer_type in ('segmentation'))
+
+    with Storage(self.layer_cloudpath) as storage:
+      storage.put_files(uploads, content_type=content_type, compress=compress)
+
+  def _generate_chunks(self, img, offset):
+    shape = Vec(*img.shape)[:3]
+    offset = Vec(*offset)[:3]
+
+    bounds = Bbox( offset, shape + offset)
+
+    alignment_check = bounds.round_to_chunk_size(self.underlying, self.voxel_offset)
+
+    if not np.all(alignment_check.minpt == bounds.minpt):
+      raise ValueError('Only grid aligned writes are currently supported. Got: {}, Volume Offset: {}, Alignment Check: {}'.format(
+        bounds, self.voxel_offset, alignment_check)
+      )
+
+    bounds = Bbox.clamp(bounds, self.bounds)
```

Comment:
I'm not sure I understand your question? All datasets can be broken into chunks and offset into space. The last chunk is allowed to be truncated.